File: examples/pod.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from numpy import *
import stochbb;
from matplotlib import pylab
import cProfile

#
# This example demonstrates how a divergence point can arise in a very simple stochastic
# model of some cognitive process. The divergence point is the earliest time-point at which the
# distributions of some response latency differ between two experimental conditions. Obviously,
# there is no divergence point if the two response latency distributions are analytic. For two
# non-analytic response latencies, a divergence point may exist.
#
# This example uses 3 Gamma distributions (of wich one is a shifted Gamma distribution) and the
# minimum(,) operator. For both conditions (control, experimental) there is a common stage C
# following some Gamma distribution. Under control condition, the common stage is followed by a
# second Gamma distributed stage X1 (C -> X1). Under the experimental condition, the common stage
# also triggers a second stage X2 with a shifted Gamma distribution parallel with the stage X1
# present under the control conditon. This stage is delayed by 300ms relative to the parallel stage
# X1, but completes much faster than X1 once it started. The response latency under the exp.
# condition is then the chain C -> max(X1, X2).
#

#
# Processing model
#
d = 300
C = stochbb.gamma(3,20)
X1 = stochbb.gamma(10,30)
# shifted gamma
X2 = stochbb.gamma(1,70) + d


# response latency control condition
#   is simply R = C + X1
Rc = C + X1
# and for the experimental condition
Re = C + stochbb.minimum(X1, X2)

Tmin, Tmax, N = 0, 1200, 1200;
t = linspace(Tmin, Tmax, N);
Tc = empty((N,)); Rc.density().eval(Tmin, Tmax, Tc)
Te = empty((N,)); Re.density().eval(Tmin, Tmax, Te)
TCc = empty((N,)); Rc.density().evalCDF(Tmin, Tmax, TCc)
TCe = empty((N,)); Re.density().evalCDF(Tmin, Tmax, TCe)

pylab.subplot(211)
pc, = pylab.plot(t, Tc, "b", lw=3)
pe, = pylab.plot(t, Te, "g", lw=3)
pod = pylab.axvline(d, color="r")
pylab.legend((pc, pe, pod),
             ("Control", "Experimental", "PoD"))

pylab.subplot(212)
pc, = pylab.plot(t, 1-TCc, "b", lw=3)
pe, = pylab.plot(t, 1-TCe, "g", lw=3)
pod = pylab.axvline(d, color="r")
pylab.legend((pc, pe, pod),
             ("Control", "Experimental", "PoD"))
pylab.savefig("pod.pdf")
pylab.show()

File: examples/pod.R
library(stochbb)

#
# Processing model
#
d  <- 300
C  <- gamma(3,20)
X1 <- gamma(10,30)
# Shifted gamma
X2 <- gamma(1,70) %+% d


# response latency control condition
#   is simply R = C + X1
# The resulting distribution of response latecies is an analytic function on (0, infinity).
Rc <- C %+% X1

# And for the experimental condition 
#   R = C + min(X1, X2).
# This might be considered as a two parallel proceses X1 (identical to the X1 under control
# condition) and X2 which is delayed by d=300 ms but once started, much faster than X1.
# The resulting distribution is not analytic on (0, infinity)! It is identical to the distribution
# of R under control condition (Rc) on the interval (0, d] and diverges for T>d.
Re <- C %+% minimum(X1, X2)

Tmin <- 0; Tmax <- 1200; N <- 1200;
t   <- seq(Tmin, Tmax, length.out=N);
Tc  <- array(0, c(N)); Rc$density$eval(Tmin, Tmax, Tc)
Te  <- array(0, c(N)); Re$density$eval(Tmin, Tmax, Te)
TCc <- array(0, c(N)); Rc$density$evalCDF(Tmin, Tmax, TCc)
TCe <- array(0, c(N)); Re$density$evalCDF(Tmin, Tmax, TCe)

# Check with exact sampler
sam <- new(ExactSampler, c(Rc, Re))
res <- array(0, c(10000,2))
sam$sample(res)

# PDFs
plot(t, Te, type="l", lty=2)
lines(t, Tc, lty=1)
abline(v=d, lty=3)

# emp. PDFs (10000 sample)
plot(density(res[,2]), lty=2)
lines(density(res[,1]), lty=1)
abline(v=d, lty=3)

# Survival plots
plot(t, 1-TCc, type="l")
lines(t, 1-TCe, lty=2)
abline(v=d, lty=3)

# uncomment to plot "zoom" to divergence point.
#N <- 100000; dt <- (Tmax-Tmin)/N
#n0 <- (d-1-Tmin)/dt; n1 <- (d+1-Tmin)/dt
#t   <- seq(Tmin, Tmax, length.out=N);
#Tc  <- array(0, c(N)); Rc$density$eval(Tmin, Tmax, Tc)
#Te  <- array(0, c(N)); Re$density$eval(Tmin, Tmax, Te)
#plot(t[n0:n1], log(Te[n0:n1]/Tc[n0:n1]), type="l")
#abline(v=d, lty=3)
#abline(h=0, lty=4)
